File: crates/hyper_render/src/renderer.rs
//
// Copyright (c) 2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::sync::Arc;

use hyper_math::Vec4;
use hyper_rhi::{
    buffer::{Buffer, BufferDescriptor, BufferUsage},
    command_list::CommandList,
    graphics_device::GraphicsDevice,
    graphics_pipeline::{GraphicsPipeline, GraphicsPipelineDescriptor},
    pipeline_layout::{PipelineLayout, PipelineLayoutDescriptor},
    render_pass::{DrawIndexedArguments, RenderPassDescriptor},
    shader_module::{ShaderModuleDescriptor, ShaderStage},
    surface::Surface,
};

use crate::{material::Material, mesh::Mesh, push_constants::ObjectPushConstants, scene::Scene};

pub struct Renderer {
    frame_index: u32,

    indices: Arc<dyn Buffer>,
    mesh: Arc<dyn Buffer>,
    _normals: Arc<dyn Buffer>,
    _positions: Arc<dyn Buffer>,
    material: Arc<dyn Buffer>,

    pipeline: Arc<dyn GraphicsPipeline>,
    _pipeline_layout: Arc<dyn PipelineLayout>,

    command_list: Arc<dyn CommandList>,

    graphics_device: Arc<dyn GraphicsDevice>,
}

impl Renderer {
    pub fn new(graphics_device: &Arc<dyn GraphicsDevice>) -> Self {
        let command_list = graphics_device.create_command_list();

        let pipeline_layout = graphics_device.create_pipeline_layout(&PipelineLayoutDescriptor {
            label: Some("Object Pipeline Layout"),
            push_constants_size: size_of::<ObjectPushConstants>(),
        });

        let vertex_shader = graphics_device.create_shader_module(&ShaderModuleDescriptor {
            label: Some("Opaque Vertex Shader"),
            path: "./assets/shaders/opaque_shader.hlsl",
            entry_point: "vs_main",
            stage: ShaderStage::Vertex,
        });

        let fragment_shader = graphics_device.create_shader_module(&ShaderModuleDescriptor {
            label: Some("Opaque Fragment Shader"),
            path: "./assets/shaders/opaque_shader.hlsl",
            entry_point: "fs_main",
            stage: ShaderStage::Fragment,
        });

        let pipeline = graphics_device.create_graphics_pipeline(&GraphicsPipelineDescriptor {
            label: Some("Opaque Graphics Pipeline"),
            layout: &pipeline_layout,
            vertex_shader: &vertex_shader,
            fragment_shader: &fragment_shader,
        });

        let material = graphics_device.create_buffer(&BufferDescriptor {
            label: Some("Material Buffer"),
            data: bytemuck::cast_slice(&[Material {
                base_color: Vec4::new(1.0, 0.0, 0.0, 1.0),
            }]),
            usage: BufferUsage::STORAGE,
        });

        let positions = graphics_device.create_buffer(&BufferDescriptor {
            label: Some("Positions Buffer"),
            data: bytemuck::cast_slice(&[
                Vec4::new(-0.5, -0.5, 0.0, 1.0),
                Vec4::new(0.5, -0.5, 0.0, 1.0),
                Vec4::new(0.5, 0.5, 0.0, 1.0),
                Vec4::new(-0.5, 0.5, 0.0, 1.0),
            ]),
            usage: BufferUsage::STORAGE,
        });

        let normals = graphics_device.create_buffer(&BufferDescriptor {
            label: Some("Normals Buffer"),
            data: bytemuck::cast_slice(&[Vec4::ZERO, Vec4::ZERO, Vec4::ZERO, Vec4::ZERO]),
            usage: BufferUsage::STORAGE,
        });

        let mesh = graphics_device.create_buffer(&BufferDescriptor {
            label: Some("Mesh Buffer"),
            data: bytemuck::cast_slice(&[Mesh {
                positions: positions.handle(),
                normals: normals.handle(),
                ..Default::default()
            }]),
            usage: BufferUsage::STORAGE,
        });

        let indices = graphics_device.create_buffer(&BufferDescriptor {
            label: Some("Index Buffer"),
            data: bytemuck::cast_slice(&[0_u32, 1_u32, 2_u32, 2_u32, 3_u32, 0_u32, 0, 0]),
            usage: BufferUsage::INDEX,
        });

        Self {
            frame_index: 1,

            indices,
            mesh,
            _normals: normals,
            _positions: positions,
            material,

            pipeline,
            _pipeline_layout: pipeline_layout,

            command_list,

            graphics_device: Arc::clone(graphics_device),
        }
    }

    pub fn render(&mut self, mut surface: &mut Box<dyn Surface>, _scene: &Scene) {
        self.graphics_device
            .begin_frame(&mut surface, self.frame_index);

        let swapchain_texture = surface.current_texture();

        self.command_list.begin();

        {
            let render_pass = self.command_list.begin_render_pass(&RenderPassDescriptor {
                label: Some("Main Render Pass"),
                color_attachment: &swapchain_texture,
                depth_stencil_attachment: None,
            });

            render_pass.set_pipeline(&self.pipeline);
            render_pass.set_index_buffer(&self.indices);
            render_pass.set_push_constants(bytemuck::bytes_of(&ObjectPushConstants {
                mesh: self.mesh.handle(),
                material: self.material.handle(),
                ..Default::default()
            }));

            render_pass.draw_indexed(&DrawIndexedArguments {
                index_count: 6,
                instance_count: 1,
                first_index: 0,
                vertex_offset: 0,
                first_instance: 0,
            });
        }

        self.command_list.end();

        self.graphics_device.execute(&self.command_list);
        self.graphics_device.present(surface);

        self.frame_index += 1;
    }

    pub fn render_ui(&self) {}
}

File: hyper_render/src/hyper_render/renderer.cpp
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_render/renderer.hpp"

#include <array>

#include <glm/glm.hpp>

#include <hyper_core/filesystem.hpp>
#include <hyper_core/logger.hpp>

struct Material
{
    glm::vec4 base_color;
};

struct Mesh
{
    uint32_t positions;
    uint32_t normals;
    uint32_t padding_0;
    uint32_t padding_1;
};

struct ObjectPushConstants
{
    uint32_t mesh;
    uint32_t material;
    uint32_t padding_0;
    uint32_t padding_1;
};

namespace hyper_render
{
    static constexpr std::array<Material, 1> s_materials = {
        Material{
            .base_color = glm::vec4(1.0f, 0.0f, 0.0f, 1.0f),
        },
    };

    static constexpr std::array<glm::vec4, 4> s_positions = {
        glm::vec4(-0.5, -0.5, 0.0, 1.0),
        glm::vec4(0.5, -0.5, 0.0, 1.0),
        glm::vec4(0.5, 0.5, 0.0, 1.0),
        glm::vec4(-0.5, 0.5, 0.0, 1.0),
    };

    static constexpr std::array<glm::vec4, 4> s_normals = {
        glm::vec4(0.0, 0.0, 0.0, 0.0),
        glm::vec4(0.0, 0.0, 0.0, 0.0),
        glm::vec4(0.0, 0.0, 0.0, 0.0),
        glm::vec4(0.0, 0.0, 0.0, 0.0),
    };

    static constexpr std::array<uint32_t, 6> s_indices = {
        0, 1, 2, 2, 3, 0,
    };

    Renderer::Renderer(const RendererDescriptor &descriptor)
        : m_graphics_device(descriptor.graphics_device)
        , m_surface(descriptor.surface)
        , m_command_list(m_graphics_device->create_command_list())
        , m_pipeline_layout(m_graphics_device->create_pipeline_layout({
              .label = "Opaque Pipeline Layout",
              .push_constant_size = sizeof(ObjectPushConstants),
          }))
        , m_vertex_shader(m_graphics_device->create_shader_module({
              .label = "Opaque Vertex Shader",
              .type = hyper_rhi::ShaderType::Vertex,
              .entry_name = "vs_main",
              .bytes = hyper_core::filesystem::read_file("./assets/shaders/opaque_shaders.hlsl"),
          }))
        , m_fragment_shader(m_graphics_device->create_shader_module({
              .label = "Opaque Fragment Shader",
              .type = hyper_rhi::ShaderType::Fragment,
              .entry_name = "fs_main",
              .bytes = hyper_core::filesystem::read_file("./assets/shaders/opaque_shaders.hlsl"),
          }))
        , m_pipeline(m_graphics_device->create_graphics_pipeline({
              .label = "Opaque Pipeline",
              .layout = m_pipeline_layout,
              .vertex_shader = m_vertex_shader,
              .fragment_shader = m_fragment_shader,
          }))
        , m_material_buffer(m_graphics_device->create_buffer({
              .label = "Material Buffer",
              .byte_size = sizeof(s_materials),
              .is_index_buffer = false,
              .is_constant_buffer = true,
          }))
        , m_positions_buffer(m_graphics_device->create_buffer({
              .label = "Positions Buffer",
              .byte_size = sizeof(s_positions),
              .is_index_buffer = false,
              .is_constant_buffer = true,
          }))
        , m_normals_buffer(m_graphics_device->create_buffer({
              .label = "Normals Buffer",
              .byte_size = sizeof(s_normals),
              .is_index_buffer = false,
              .is_constant_buffer = true,
          }))
        , m_mesh_buffer(m_graphics_device->create_buffer({
              .label = "Mesh Buffer",
              .byte_size = sizeof(Mesh) * 1,
              .is_index_buffer = false,
              .is_constant_buffer = true,
          }))
        , m_indices_buffer(m_graphics_device->create_buffer({
              .label = "Indices Buffer",
              .byte_size = sizeof(s_indices),
              .is_index_buffer = true,
              .is_constant_buffer = false,
          }))
        , m_frame_index(1)
    {
        m_graphics_device->wait_for_idle();

        HE_DEBUG("Created Renderer");
    }

    void Renderer::render()
    {
        m_graphics_device->begin_frame(m_surface, m_frame_index);

        m_command_list->begin();

        m_command_list->end();

        m_graphics_device->end_frame();

        m_graphics_device->execute();
        m_graphics_device->present(m_surface);

        m_frame_index += 1;
    }
} // namespace hyper_render